File: 2024-12-December-GFG-POTD/December 01 - Non Repeating Character/Solution.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    // Function to find the first non-repeating character in a string.
    char nonRepeatingChar(string s) {
        // Step 1: Create a frequency map to count occurrences of each character
        unordered_map<char, int> freqMap;
        
        // Step 2: Populate the frequency map
        for (char c : s) {
            freqMap[c]++;
        }
        
        // Step 3: Traverse the string again to find the first non-repeating character
        for (char c : s) {
            if (freqMap[c] == 1) {
                return c; // Return the first non-repeating character
            }
        }
        
        // Step 4: If no non-repeating character is found, return '$'
        return '$';
    }
};

int main() {
    int t;
    cin >> t;
    
    while (t--) {
        string s;
        cin >> s;
        
        Solution obj;
        char ans = obj.nonRepeatingChar(s);
        
        if (ans != '$')
            cout << ans << endl;
        else
            cout << -1 << endl;

        cout << "~" << endl;
    }
    
    return 0;
}

File: 2024-12-December-GFG-POTD/December 01 - Non Repeating Character/Solution.py
class Solution:
    # Function to find the first non-repeating character in a string.
    def nonRepeatingChar(self, s: str) -> str:
        # Step 1: Create a frequency map to count occurrences of each character
        freq_map = {}
        
        # Step 2: Populate the frequency map
        for c in s:
            freq_map[c] = freq_map.get(c, 0) + 1
        
        # Step 3: Traverse the string again to find the first non-repeating character
        for c in s:
            if freq_map[c] == 1:
                return c  # Return the first non-repeating character
        
        # Step 4: If no non-repeating character is found, return '$'
        return '$'

# Driver code
if __name__ == "__main__":
    t = int(input())
    
    for _ in range(t):
        s = input()
        
        obj = Solution()
        ans = obj.nonRepeatingChar(s)
        
        if ans != '$':
            print(ans)
        else:
            print(-1)
        print("~")
